CCGOutcomeRuleTest fails

The Test for the CCGOutcomeRule sometimes fails with a java.util.ConcurrentModificationException. You can find the complete stack trace below:

`java.util.ConcurrentModificationException: java.util.ConcurrentModificationException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:600)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)
	at java.base/java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:722)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at org.marketdesignresearch.mechlib.outcomerules.ccg.constraintgeneration.ValueSeparabilityGenerator.generateConstraint(ValueSeparabilityGenerator.java:65)
	at org.marketdesignresearch.mechlib.outcomerules.ccg.constraintgeneration.UnitedConstrainedGenerator.addConstraint(UnitedConstrainedGenerator.java:60)
	at org.marketdesignresearch.mechlib.outcomerules.ccg.CCGOutcomeRule.calculateCCGPrices(CCGOutcomeRule.java:72)
	at org.marketdesignresearch.mechlib.outcomerules.ccg.CCGOutcomeRule.getOutcome(CCGOutcomeRule.java:53)
	at org.marketdesignresearch.mechlib.outcomerules.OutcomeRule.getPayment(OutcomeRule.java:12)
	at org.marketdesignresearch.mechlib.outcomerules.ccg.CCGOutcomeRuleTest.smallMultiConstraint(CCGOutcomeRuleTest.java:225)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:542)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:770)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: java.util.ConcurrentModificationException
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1134)
	at org.jgrapht.alg.util.NeighborCache.fetch(NeighborCache.java:122)
	at org.jgrapht.alg.util.NeighborCache.neighborsOf(NeighborCache.java:96)
	at org.marketdesignresearch.mechlib.outcomerules.ccg.constraintgeneration.ValueSeparabilityGenerator.createConflictGraph(ValueSeparabilityGenerator.java:87)
	at org.marketdesignresearch.mechlib.outcomerules.ccg.constraintgeneration.ValueSeparabilitySideDownGenerator.investigateSubCoalition(ValueSeparabilitySideDownGenerator.java:26)
	at org.marketdesignresearch.mechlib.outcomerules.ccg.constraintgeneration.ValueSeparabilityGenerator.lambda$2(ValueSeparabilityGenerator.java:74)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:408)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:736)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:661)
	at org.marketdesignresearch.mechlib.outcomerules.ccg.constraintgeneration.ValueSeparabilityGenerator.addConstraints(ValueSeparabilityGenerator.java:75)
	at org.marketdesignresearch.mechlib.outcomerules.ccg.constraintgeneration.ValueSeparabilityGenerator.lambda$0(ValueSeparabilityGenerator.java:61)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)`

--- a/src/main/java/org/marketdesignresearch/mechlib/outcomerules/ccg/constraintgeneration/ValueSeparabilityGenerator.java
+++ b/src/main/java/org/marketdesignresearch/mechlib/outcomerules/ccg/constraintgeneration/ValueSeparabilityGenerator.java
@@ -1,12 +1,7 @@
 package org.marketdesignresearch.mechlib.outcomerules.ccg.constraintgeneration;
 
 import java.math.BigDecimal;
-import java.util.LinkedList;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Set;
-import java.util.concurrent.ForkJoinTask;
+import java.util.*;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Consumer;
 
@@ -52,17 +47,15 @@ public abstract class ValueSeparabilityGenerator implements PartialConstraintGen
                                    ConnectivityInspector<PotentialCoalition, DefaultEdge> connectivityInspector, Allocation blockingCoalition, Outcome priorResult) {
         constraintCount.set(0);
         NeighborCache<PotentialCoalition, DefaultEdge> neighbors = new NeighborCache<>(graph);
-        List<ForkJoinTask<?>> graphInspecting = new LinkedList<>();
         for (Set<PotentialCoalition> subgraph : connectivityInspector.connectedSets()) {
-            ForkJoinTask<?> task = ForkJoinTask.adapt(() -> {
-                BlockedBidders blockedBidders = BlockedBidders.from(subgraph, blockingCoalition.getPotentialCoalitions());
-                AverageDistanceFromReference distancePerBidder = calcAdr(blockedBidders, priorResult.getPayment(), blockedBidders.getNonTraitors());
-
-                addConstraints(blockedBidders, neighbors, priorResult.getPayment(), corePaymentRule, distancePerBidder);
-            });
-            graphInspecting.add(task.fork());
+            // Compared to a previous implementation, the parallel aspect was removed here. The problem was that
+            // jgrapht's neighborsOf() method threw a ConcurrentModificationException in a computeIfAbsent() call
+            // when parallelized. This was possible due to a bug in JDK8 (which is why it remained undetected), but not
+            // in later JDKs
+            BlockedBidders blockedBidders = BlockedBidders.from(subgraph, blockingCoalition.getPotentialCoalitions());
+            AverageDistanceFromReference distancePerBidder = calcAdr(blockedBidders, priorResult.getPayment(), blockedBidders.getNonTraitors());
+            addConstraints(blockedBidders, neighbors, priorResult.getPayment(), corePaymentRule, distancePerBidder);
         }
-        graphInspecting.forEach(ForkJoinTask::join);
     }
 
     protected void addConstraints(BlockedBidders blockedBidders, NeighborCache<PotentialCoalition, DefaultEdge> neighbors, Payment priorPayment, CorePaymentRule corePaymentRule,
@@ -71,8 +64,9 @@ public abstract class ValueSeparabilityGenerator implements PartialConstraintGen
             corePaymentRule.addBlockingConstraint(blockedBidders, priorPayment);
 
             if (constraintCount.getAndIncrement() < MAX_CONSTRAINTS) {
-                Consumer<Bidder> subCoalitionInvestigator = b -> investigateSubCoalition(blockedBidders.getBlockingBidders(), neighbors, b, minadr, priorPayment, corePaymentRule);
-                blockedBidders.getBlockedBidders().parallelStream().forEach(subCoalitionInvestigator);
+                for (Bidder blockedBidder : blockedBidders.getBlockedBidders()) {
+                    investigateSubCoalition(blockedBidders.getBlockingBidders(), neighbors, blockedBidder, minadr, priorPayment, corePaymentRule);
+                }
             }
         }
 

--- a/src/main/java/org/marketdesignresearch/mechlib/outcomerules/ccg/constraintgeneration/ValueSeparabilityGenerator.java
+++ b/src/main/java/org/marketdesignresearch/mechlib/outcomerules/ccg/constraintgeneration/ValueSeparabilityGenerator.java
@@ -51,7 +51,7 @@ public abstract class ValueSeparabilityGenerator implements PartialConstraintGen
             // Compared to a previous implementation, the parallel aspect was removed here. The problem was that
             // jgrapht's neighborsOf() method threw a ConcurrentModificationException in a computeIfAbsent() call
             // when parallelized. This was possible due to a bug in JDK8 (which is why it remained undetected), but not
-            // in later JDKs
+            // in later JDKs: https://stackoverflow.com/a/54825115
             BlockedBidders blockedBidders = BlockedBidders.from(subgraph, blockingCoalition.getPotentialCoalitions());
             AverageDistanceFromReference distancePerBidder = calcAdr(blockedBidders, priorResult.getPayment(), blockedBidders.getNonTraitors());
             addConstraints(blockedBidders, neighbors, priorResult.getPayment(), corePaymentRule, distancePerBidder);

